### PR TITLE
Class names to enable isinstance

### PR DESCRIPTION
**Pull request by dschult**
Add empty class sparray to allow ID of array api.

~~Change xxx_array to _xxx_array and add empty class to allow ID of xxx_array.~~
~~Add functions is_xxx_array, is_xxx_matrix and is_xxx~~
Note: Can check for is_xxx using isinstance(A, _xxx_array) or isinstance(A, (xxx_array, xxx_matrix)).

~~Add functions issparray, is_xxx, is_xxx_array, is_xxx_matrix~~

Questions:
- Naming conventions? is_xxx(A), is_xxx_array(A),  {probs keep isspmatrix_xxx(A) right?}  But we have a chance to choose for sparrays.
- Naming conventions? issparray(A) or is_sparray(A)?  {Keep isspmatrix, right?}
- we can check is_xxx using `A._format == 'xxx'` or using `isinstance(A, _xxx_array)`. Which is better?  (code shows both for now)
- we can check for issparray using `A._is_array` or using `isinstance(A, sparray)`. Which is better?  (code shows both for now)

[Edited]
Answers:
- don't pollute with lots of new "is_" functions. Let users use `A.format` and `A._is_array` if they need to check these.
- do make the class hierarchy so that `isinstance` makes it clear how to test format and array/matrix.

```
issparse(A)            <=>     isinstance(A, _sparray)           (2nd option isn't advertised)
isspmatrix(A)         <=>     isinstance(A, spmatrix)           (as before -- backward compat)
isinstance(A, csr_matrix)                                                    (as before -- backward compat)

isinstance(A, sparray)
isinstance(A, csr_array)

isinstance(A, (csr_matrix, csr_array))    OR  isinstance(A, _csr_array) but 2nd option isn't advertised
```

### CODE REVIEW

File: scipy/sparse/_bsr.py [hunk 1]
```diff
@@ -722,7 +722,10 @@ def isspmatrix_bsr(x):
     return isinstance(x, bsr_matrix) or isinstance(x, bsr_array)
 
 
-class bsr_matrix(spmatrix, bsr_array):
+class bsr_array(_bsr_array, sparray):
```
Comment (jjerphan) [inline scipy/sparse/_bsr.py:725]: Naive question: must not we use the same method resolution order for sparse arrays and sparse matrices?

**Comment by stefanv** [comment]
The current namings are hard to disambiguate: `_sparray` vs `sparray`, `_csr_array` vs `csr_array`. After discussion, decided to rename `_sparray` to `_spbase` (this is the base class for both array and matrix implementations). Similarly, `_<format>_array` will become `_<format>_base` (or `_<format>_impl` or similar), with `<format>_array` and `<format>_matrix` subclassing that, and mixing in either `spmatrix` or `sparray` to help with isinstance checking. `issparse` will only distinguish between dense and sparse objects, and we will not implement `issparray` and `isspmatrix` (use isinstance checks instead).

**Comment by dschult** [comment]
There is a related discussion in [#15790](https://github.com/scipy/scipy/issues/15790#issuecomment-1069330330) describing these  naming changes.

File: scipy/sparse/_bsr.py (hunk 1 above)
Comment (dschult) [reply scipy/sparse/_bsr.py:725]: The order matters for method lookup order. For `bsr_matrix`, presumably there are methods in both classes so the order is important. 

For `bsr_array` the `sparray` class is an empty class (used for it's isinstance impact to indicate an array class). So order doesn't impact the result, and being last might have a very slight speedup for method lookup -- the methods are all in the `_bsr_array` class.

Comment (jjerphan) [reply scipy/sparse/_bsr.py]: Do we want to have the same order (e.g. extend the format mixing first and `sp{array,matrix}` second) for consistency?

File: scipy/sparse/_csr.py [hunk 2]
```diff
@@ -131,7 +131,7 @@ class csr_array(_cs_matrix):
            [0, 1, 1, 1]])
 
     """
-    _format = 'csr'
+    format = 'csr'
```
Comment (perimosocordiae) [inline scipy/sparse/_csr.py:134]: This should be reverted

**Review by perimosocordiae — APPROVED** [review]

File: scipy/sparse/_bsr.py [hunk 3]
```diff
@@ -722,7 +722,11 @@ def isspmatrix_bsr(x):
     return isinstance(x, bsr_matrix) or isinstance(x, bsr_array)
 
 
-class bsr_matrix(spmatrix, bsr_array):
+# This namespace class separates array from matrix with ininstance
```
Comment (rossbar) [inline scipy/sparse/_bsr.py:725]:
```suggestion
# This namespace class separates array from matrix with isinstance
```

File: scipy/sparse/_coo.py [hunk 4]
```diff
@@ -620,7 +620,11 @@ def isspmatrix_coo(x):
     return isinstance(x, coo_matrix) or isinstance(x, coo_array)
 
 
-class coo_matrix(spmatrix, coo_array):
+# This namespace class separates array from matrix with ininstance
```
Comment (rossbar) [inline scipy/sparse/_coo.py:623]:
```suggestion
# This namespace class separates array from matrix with isinstance
```

File: scipy/sparse/_csc.py [hunk 5]
```diff
@@ -260,7 +260,11 @@ def isspmatrix_csc(x):
     return isinstance(x, csc_matrix) or isinstance(x, csc_array)
 
 
-class csc_matrix(spmatrix, csc_array):
+# This namespace class separates array from matrix with ininstance
```
Comment (rossbar) [inline scipy/sparse/_csc.py:263]:
```suggestion
# This namespace class separates array from matrix with isinstance
```

File: scipy/sparse/_csr.py [hunk 6]
```diff
@@ -357,7 +357,11 @@ def isspmatrix_csr(x):
     return isinstance(x, csr_matrix) or isinstance(x, csr_array)
 
 
-class csr_matrix(spmatrix, csr_array):
+# This namespace class separates array from matrix with ininstance
```
Comment (rossbar) [inline scipy/sparse/_csr.py:360]:
```suggestion
# This namespace class separates array from matrix with isinstance
```

File: scipy/sparse/_dia.py [hunk 7]
```diff
@@ -469,7 +469,11 @@ def isspmatrix_dia(x):
     return isinstance(x, dia_matrix) or isinstance(x, dia_array)
 
 
-class dia_matrix(spmatrix, dia_array):
+# This namespace class separates array from matrix with ininstance
```
Comment (rossbar) [inline scipy/sparse/_dia.py:472]:
```suggestion
# This namespace class separates array from matrix with isinstance
```

File: scipy/sparse/_dok.py [hunk 8]
```diff
@@ -448,7 +448,11 @@ def isspmatrix_dok(x):
     return isinstance(x, dok_matrix) or isinstance(x, dok_array)
 
 
-class dok_matrix(spmatrix, dok_array):
+# This namespace class separates array from matrix with ininstance
```
Comment (rossbar) [inline scipy/sparse/_dok.py:451]:
```suggestion
# This namespace class separates array from matrix with isinstance
```

File: scipy/sparse/_lil.py [hunk 9]
```diff
@@ -547,7 +547,11 @@ def isspmatrix_lil(x):
     return isinstance(x, lil_matrix) or isinstance(x, lil_array)
 
 
-class lil_matrix(spmatrix, lil_array):
+# This namespace class separates array from matrix with ininstance
```
Comment (rossbar) [inline scipy/sparse/_lil.py:550]:
```suggestion
# This namespace class separates array from matrix with isinstance
```

**Review by rossbar — APPROVED** [review]
Thanks @dschult ! Just one nit about a typo (repeated several times)

**Comment by perimosocordiae** [comment]
The failing CI is unrelated (known issue in scipy.optimize tests). This is ready to merge.

**Pull request merged**